Add more robust support for line number, column number and byte number tracking to the lexer

neuroevolutus opened the pull request:
The first batch of commits in this pull request involve minor refactors that were found to be potentially worthwhile in the course of development.

The latter half of this PR adds code in service of eventually closing #86. The main additions are the classes `SourceLocation` and `SourceExtent`. `class SourceLocation` is a dataclass that stores a line number, column number, and byte index into the scrapscript program source.

`class SourceExtent` has `start` and `end` members of type `SourceLocation` to represent a swathe of contiguous source code from the user's scrapscript program. The line and column numbers are 1-indexed as before whereas the byte numbers are 0-indexed. All end indices are inclusive of the final characters of their corresponding tokens, i.e. they do *not* demarcate the positions of "one past the end" of the tokens.

The classes `Token` and `Lexer` are modified to work in terms of `SourceExtent` so that the lexing phase can maintain richer source information. This new code structure will hopefully be more robust and hopefully allow for easier refactoring in the future if needed.

I expect to do a follow-up PR that can build atop this one to maintain richer source information tracking for the parsing stage.

neuroevolutus committed:
Rename function `read_one` to `read_token` to improve clarity

neuroevolutus committed:
Remove unused `pytest` dependency

Poetry's deprecated `category` key is also replaced with the key `groups`.

neuroevolutus committed:
Add `num_bytes_as_utf8` function

neuroevolutus committed:
Add `class UnexpectedTokenError`

neuroevolutus committed:
Raise `UnexpectedTokenError` in place of more general `ParseError`

neuroevolutus committed:
Add comment

neuroevolutus committed:
Alter `Lexer.read_bytes` to use `peek_char`

This will make it more amenable to source location tracking.

neuroevolutus committed:
Modify lexer to utilize `class SourceExtent`

`class Token` now takes a `SourceExtent` rather than an `int` value.

Instances of variables having `lineno` and `colno` in their names are
altered to have `line_number` and `column_number` in their spellings
instead.

Notice that `Lexer._line_number`, `Lexer._column_number` and
`Lexer._byte_number` are made to be unmodifiable members using the
`@property` decorator.

neuroevolutus committed:
Update variant tokenization test to test more malicious whitespace

neuroevolutus committed:
Rename some tests and local variables to reflect new naming scheme

`lineno` -> `line_number`
`colno` -> `column_number`

neuroevolutus committed:
Update old test to check for `SourceExtent` tracking

neuroevolutus committed:
Add `SourceExtent` tracking tests for various token types

neuroevolutus committed:
Add `class SyntacticError`

neuroevolutus committed:
Make `class ParseError` inherit from `class SyntacticError`

This will remove some dependency on the Python standard library.

neuroevolutus committed:
Add `test_read_char_increments_byte_number`

neuroevolutus committed:
Add `class SourceLocation`

neuroevolutus committed:
Add `class SourceExtent`

neuroevolutus committed:
Fix tests that must expect `UnexpectedTokenError`s

tekknolagi commented:
Wow! This is really excellent work. Thank you for the commits. I will take some time to go through them.

tekknolagi commented on scrapscript.py:
```diff
@@ -1838,8 +2061,8 @@ def test_tokenize_record_with_spread_no_spaces(self) -> None:
 
-    def test_tokenize_variant_with_space(self) -> None:
-        self.assertEqual(tokenize("# abc"), [VariantToken("abc")])
+    def test_tokenize_variant_with_whitespace(self) -> None:
```
I think some of the `VariantToken` stuff is a holdover from an earlier implementation of variants that kind of got them wrong and I think this makes some of the lexer stuff tricky. Perhaps we can refactor so the lexer instead produces `#` and `abc` as separate tokens and then the parser can combine them? As far as I'm aware, nothing else uses `#`. This might remove the need for `currently_reading_variant_token` and its associated state complexity.

neuroevolutus replied:
Oh, that would be nice. Yeah, let me take a look at that again.

tekknolagi commented on scrapscript.py:
```diff
@@ -1715,31 +1801,40 @@ def test_tokenize_compose_reverse(self) -> None:
+    def test_first_line_number_is_one(self) -> None:
         l = Lexer("abc")
-        self.assertEqual(l.lineno, 1)
+        self.assertEqual(l.line_number, 1)
```
Minor nit, but I think for a change as big as this one this kind of variable name change looks like line noise. Yes, `column_number` is a little bit clearer, but that is not the best part of your series of commits. Perhaps that can come later or before. I just like making the smallest possible precise diff. I don't feel super strongly about this for your PR though.

neuroevolutus replied:
Sure thing. I made a commit that undoes these changes.

tekknolagi commented on scrapscript.py:
```diff
@@ -252,17 +329,17 @@ def read_var(self, first_char: str) -> Token:
         buf = ""
         while self.has_input():
-            if (c := self.read_char()).isspace():
+            if (self.peek_char()).isspace():
```
Nit: drop parens around method call

neuroevolutus replied:
Oh, lol. Thanks for catching that, Max.

tekknolagi commented on scrapscript.py:
```diff
@@ -191,7 +268,7 @@ def read_one(self) -> Token:
         if is_identifier_char(c):
             return self.read_var(c)
-        raise ParseError(f"unexpected token {c!r}", ("<input>", self.lineno, self.colno, self.line))
+        raise ParseError(f"unexpected token {c!r}", ("<input>", self._line_number, self._column_number, self.line))
```
Use `UnexpectedTokenError`?

neuroevolutus replied:
I tried introducing a new `class InvalidTokenError` for representing sequences of source characters that do not correspond to any valid token. My thinking was that `class UnexpectedTokenError` could be used to indicate that the parser found a valid lexical token that was out of place grammatically. Let me know what you think; I'll be happy to change this if this is splitting hairs too much.

tekknolagi commented on scrapscript.py:
```diff
@@ -161,10 +234,14 @@ def read_one(self) -> Token:
                 self.read_comment()
-                return self.read_one()
+                # Need to start reading a new token
+                self.mark_token_start()
```
If we drop the variant shenanigans, does that mean we can have a clean entry of `mark_token_start` at the top of `read_token`?

neuroevolutus replied:
Indeed! Yes, you are right.

tekknolagi commented on scrapscript.py:
```diff
@@ -252,17 +329,17 @@ def read_var(self, first_char: str) -> Token:
+            if (self.peek_char()).isspace():
                 break
-            buf += c
+            buf += self.read_char()
```
Can we revert this bit if we fix the variant issues?

neuroevolutus replied:
With regards to this and the question about `make_token`, I think I might be able to get rid of this circuitous logic if `SourceExtent.end` were to track the _exclusive_ end of a token within the source. Thus far, I followed the convention of tracking the _inclusive_ ends of tokens within the source and so had to be careful not to lose my proper position within the text by chomping away characters too early. Let me know if you'd like me to alter this.

tekknolagi replied:
I think we can keep as is for now and change later if needed

tekknolagi commented on scrapscript.py:
```diff
@@ -146,11 +205,25 @@ def peek_char(self) -> str:
 
     def make_token(self, cls: type, *args: Any) -> Token:
```
Why don't we mark token end in `make_token`? I think I am missing something

tekknolagi reviewed (commented):
I really like these changes. I think we're very close to mergeable. If you're amenable, I have a couple of suggestions.

Additionally: what does using this API look like in the parser? I think seeing even a fictitious example of that might clarify some other questions I have.

Thank you!!!

tekknolagi commented:
A lot of it is just: in general, I like "surgical" changes in a PR. Among other things, it makes the commit history and blame cleaner

neuroevolutus committed:
Remove redundant pair of parentheses

neuroevolutus committed:
Add `class Hash`

neuroevolutus committed:
Remove `class VariantToken`

Variants are now lexed separately into two tokens, i.e. `Hash` and
`Name`.

neuroevolutus committed:
Fix tests that rely on defunct `class VariantToken`

neuroevolutus committed:
Add new parser tests to account for updated variant lexing

neuroevolutus committed:
Add `class InvalidTokenError`

This should be raised when strings of characters that do not
correspond to any valid token are encountered.

neuroevolutus committed:
Remove unnecessary use of `Lexer.mark_token_start()`

neuroevolutus committed:
Exchange a couple of lines to make them match the comments

neuroevolutus committed:
Raise `InvalidTokenError` when lexer finds invalid source characters

neuroevolutus committed:
Update naming scheme for source extents to utilize old convention

`line_number` -> `lineno`
`column_number` -> `colno`
`byte_number` -> `byteno`

neuroevolutus commented:
> I really like these changes. I think we're very close to mergeable. If you're amenable, I have a couple of suggestions.
> 
> Additionally: what does using this API look like in the parser? I think seeing even a fictitious example of that might clarify some other questions I have.
> 
> Thank you!!!

Thanks for the feedback so far, Max. What I was thinking for the parser was along the following lines.

I'm imagining that AST nodes will also have a member variable of type `SourceExtent`. When we're in the parser and aggregating two or more tokens into a single AST node, the `start` of its `SourceExtent` would take on the minimum value among the values of `SourceExtent.start` of its component tokens. Similarly, the `end` of its `SourceExtent` would take on the maximum among the values of `SourceExtent.end` of its component tokens.

Working our way up from the leaves to the root of the AST, we would find that the value of `start` would eventually represent the start of the source file and `end` would eventually represent the end of the source file.

With regards to handling interstitial values as you mentioned in #86, perhaps we could make this `source_extent` member of the AST node an `Optional` so that not every node is required to have a `SourceExtent`.

tekknolagi reviewed (approved):
Nice! Love it.

tekknolagi commented:
This is very exciting. Merged. Excited to see the follow up parser work.

tekknolagi commented:
Thank you!!
